File: client/csharp/Angzarr.Client/Router/ProjectorRouter.cs
namespace Angzarr.Client.Router;

/// <summary>
/// Router for projector components (events -> external output, multi-domain).
///
/// Domains are registered via fluent Domain() calls.
///
/// Example:
/// <code>
/// var router = new ProjectorRouter("prj-output")
///     .Domain("player", new PlayerProjectorHandler())
///     .Domain("hand", new HandProjectorHandler());
///
/// // Dispatch events
/// var projection = router.Dispatch(eventBook);
/// </code>
/// </summary>
public class ProjectorRouter
{
    private readonly string _name;
    private readonly Dictionary<string, IProjectorDomainHandler> _domains = new();

    /// <summary>
    /// Create a new projector router.
    /// Projectors consume events from multiple domains and produce external output.
    /// </summary>
    /// <param name="name">Component name.</param>
    public ProjectorRouter(string name)
    {
        _name = name;
    }

    /// <summary>
    /// Register a domain handler.
    /// Projectors can have multiple input domains.
    /// </summary>
    /// <param name="name">Domain name.</param>
    /// <param name="handler">Handler for events from this domain.</param>
    /// <returns>This router for fluent chaining.</returns>
    public ProjectorRouter Domain<THandler>(string name, THandler handler)
        where THandler : IProjectorDomainHandler
    {
        _domains[name] = handler;
        return this;
    }

    /// <summary>
    /// Get the router name.
    /// </summary>
    public string Name => _name;

    /// <summary>
    /// Get subscriptions (domain + event types) for this projector.
    /// Returns list of (domain, event types) tuples.
    /// </summary>
    public IReadOnlyList<(string Domain, IReadOnlyList<string> Types)> Subscriptions()
    {
        return _domains
            .Select(kv => (kv.Key, (IReadOnlyList<string>)kv.Value.EventTypes()))
            .ToList();
    }

    /// <summary>
    /// Dispatch events to the appropriate handler.
    /// </summary>
    /// <param name="events">Event book to process.</param>
    /// <returns>Projection result.</returns>
    public Angzarr.Projection Dispatch(Angzarr.EventBook events)
    {
        var domain = events.Cover?.Domain ?? "";

        if (!_domains.TryGetValue(domain, out var handler))
            throw new InvalidArgumentError($"No handler for domain: {domain}");

        return handler.Project(events);
    }
}

File: client/java/client/src/main/java/dev/angzarr/client/router/ProjectorRouter.java
package dev.angzarr.client.router;

import dev.angzarr.*;

import java.util.*;

/**
 * Router for projector components (events -> external output, multi-domain).
 *
 * <p>Domains are registered via fluent {@code .domain()} calls, supporting
 * multiple input domains.
 *
 * <p>Example:
 * <pre>{@code
 * ProjectorRouter router = ProjectorRouter.create("prj-output")
 *     .domain("player", new PlayerProjectorHandler())
 *     .domain("hand", new HandProjectorHandler());
 *
 * // Get subscriptions for registration
 * List<Map.Entry<String, List<String>>> subs = router.subscriptions();
 *
 * // Project events
 * Projection result = router.dispatch(events);
 * }</pre>
 */
public class ProjectorRouter {

    private final String name;
    private final Map<String, ProjectorDomainHandler> domains;

    private ProjectorRouter(String name, Map<String, ProjectorDomainHandler> domains) {
        this.name = name;
        this.domains = domains;
    }

    /**
     * Create a new projector router.
     *
     * @param name The router name
     * @return A new router ready for domain registration
     */
    public static ProjectorRouter create(String name) {
        return new ProjectorRouter(name, new HashMap<>());
    }

    /**
     * Register a domain handler.
     *
     * <p>Projectors can have multiple input domains.
     * Returns a new router with the additional domain registered.
     *
     * @param domainName The domain name to listen for
     * @param handler The handler for events from this domain
     * @return A new router with the domain registered
     */
    public ProjectorRouter domain(String domainName, ProjectorDomainHandler handler) {
        Map<String, ProjectorDomainHandler> newDomains = new HashMap<>(this.domains);
        newDomains.put(domainName, handler);
        return new ProjectorRouter(name, newDomains);
    }

    /**
     * Get the router name.
     */
    public String getName() {
        return name;
    }

    /**
     * Get subscriptions (domain + event types) for this projector.
     *
     * @return List of (domain, event types) pairs
     */
    public List<Map.Entry<String, List<String>>> subscriptions() {
        List<Map.Entry<String, List<String>>> subs = new ArrayList<>();
        for (Map.Entry<String, ProjectorDomainHandler> entry : domains.entrySet()) {
            subs.add(new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue().eventTypes()));
        }
        return subs;
    }

    /**
     * Dispatch events to the appropriate handler.
     *
     * @param events The event book to project
     * @return The projection result
     * @throws RouterException if dispatch fails
     */
    public Projection dispatch(EventBook events) throws RouterException {
        String domain = events.hasCover() ? events.getCover().getDomain() : "";

        ProjectorDomainHandler handler = domains.get(domain);
        if (handler == null) {
            throw new RouterException("No handler for domain: " + domain);
        }

        try {
            return handler.project(events);
        } catch (ProjectorDomainHandler.ProjectionError e) {
            throw new RouterException("Projection failed: " + e.getMessage(), e);
        }
    }

    /**
     * Exception type for router errors.
     */
    public static class RouterException extends Exception {
        public RouterException(String message) {
            super(message);
        }

        public RouterException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}
